### PR TITLE
Fix jax.scipy.special.betaln returning nan for negative-integer inputs

### LINKED ISSUE #38243: jax.scipy.special.betaln returns nan instead of finite value for negative integer inputs
### Description

`jax.scipy.special.betaln(a, b)` returns `nan` when `a` is a negative
integer and `b <= |a|`, even though the mathematically correct result is
a finite value. SciPy correctly returns finite values in all affected cases.

The bug is systematic: for any negative integer `a = -n` and `b` with
`1 <= b <= n`, JAX returns `nan` instead of the correct finite value.

## Origin

Found via testing derived from:

```
jax/tests/lax_scipy_special_functions_test.py
  ::LaxScipySpecialFunctionsTest::testScipySpecialFun_betaln
```

The existing test uses `rand_positive` inputs only, which never include
negative values. The property uses a **reference oracle**:
`jax.scipy.special.betaln(a, b)` must agree with
`scipy.special.betaln(a, b)` for all inputs.

## Verification (SciPy is correct)

All affected cases were verified independently via **mpmath**
(arbitrary-precision library, unrelated to SciPy):

```python
import mpmath
cases = [(-1,1), (-2,1), (-2,2), (-3,1), (-3,2), (-3,3), (-4,1), (-4,4)]
for a, b in cases:
    m = float(mpmath.re(mpmath.log(abs(mpmath.beta(a, b)))))
    print(f"mpmath.betaln({a}, {b}) = {m:.4f}")
# mpmath.betaln(-1, 1) =  0.0000
# mpmath.betaln(-2, 1) = -0.6931
# mpmath.betaln(-2, 2) = -0.6931
# mpmath.betaln(-3, 1) = -1.0986
# mpmath.betaln(-3, 2) = -1.7918
# mpmath.betaln(-3, 3) = -1.0986
# mpmath.betaln(-4, 1) = -1.3863
# mpmath.betaln(-4, 4) = -1.3863
```

SciPy and mpmath agree on all cases. The bug is in JAX.

## Minimal reproducer

```python
import numpy as np
import scipy.special
import jax
import jax.scipy.special

jax.config.update("jax_enable_x64", True)

cases = [(-1, 1), (-2, 1), (-2, 2), (-3, 3), (-4, 4)]
for a, b in cases:
    a32, b32 = np.float32(a), np.float32(b)
    print(f"betaln({a}, {b}):")
    print(f"  scipy: {scipy.special.betaln(a32, b32):.4f}")                                           # finite  ← correct
    print(f"  jax:   {jax.scipy.special.betaln(a32, b32)}")                                           # nan     ← BUG
    print(f"  jit:   {jax.jit(jax.scipy.special.betaln)(jax.numpy.array(a32), jax.numpy.array(b32))}") # nan   ← BUG
```

## Expected behavior

```
betaln(-1, 1):  scipy= 0.0000, jax= 0.0000, jit= 0.0000
betaln(-2, 1):  scipy=-0.6931, jax=-0.6931, jit=-0.6931
betaln(-2, 2):  scipy=-0.6931, jax=-0.6931, jit=-0.6931
betaln(-3, 3):  scipy=-1.0986, jax=-1.0986, jit=-1.0986
betaln(-4, 4):  scipy=-1.3863, jax=-1.3863, jit=-1.3863
```

## Actual behavior

```
betaln(-1, 1):  scipy= 0.0000, jax=nan, jit=nan
betaln(-2, 1):  scipy=-0.6931, jax=nan, jit=nan
betaln(-2, 2):  scipy=-0.6931, jax=nan, jit=nan
betaln(-3, 3):  scipy=-1.0986, jax=nan, jit=nan
betaln(-4, 4):  scipy=-1.3863, jax=nan, jit=nan
```

## Additional cases

The bug affects `betaln(a, b)` when `a` is a negative integer and `b <= |a|`.
Cases where `b > |a|` are handled correctly (both JAX and SciPy return `inf`):

```
betaln(-1, 1): scipy= 0.0000, jax=nan  ← BUG
betaln(-2, 1): scipy=-0.6931, jax=nan  ← BUG
betaln(-2, 2): scipy=-0.6931, jax=nan  ← BUG
betaln(-3, 1): scipy=-1.0986, jax=nan  ← BUG
betaln(-3, 2): scipy=-1.7918, jax=nan  ← BUG
betaln(-3, 3): scipy=-1.0986, jax=nan  ← BUG
betaln(-1, 2): scipy=inf,     jax=inf  OK
betaln(-2, 3): scipy=inf,     jax=inf  OK
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:     0.10.1
jaxlib:  0.10.1
numpy:   2.4.6
scipy:   1.17.1
python:  3.12.3
OS:      Linux-6.17.0-29-generic-x86_64-with-glibc2.39
backend: cpu
```

### PR DESCRIPTION
**Pull request by RoyK0108**
## Summary

Fixes #38243.

`jax.scipy.special.betaln(a, b)` returns `nan` when `a` is a negative integer and `b <= |a|`, where the mathematically correct result is finite (and SciPy/mpmath return a finite value).

### Root cause

The implementation computes `lgamma(a) + lgamma(b) - lgamma(a + b)`. When `a` is a non-positive integer **and** `a + b` is also a non-positive integer, `gamma` has a pole at both points, so `lgamma(a)` and `lgamma(a + b)` are each `+inf` and the result is `inf - inf = nan`. The beta function is finite there because the two poles cancel.

### Fix

Taking the limit `a -> -n` via the residues of `gamma`:

```
B(a, b) = (-1)**b * gamma(b) * gamma(1 - a - b) / gamma(1 - a)
```

so in this regime `betaln(a, b) = lgamma(b) + lgamma(1 - a - b) - lgamma(1 - a)`. The fix applies this reflection formula exactly when both `a` and `a + b` are non-positive integers. As a bonus, this also fixes the case where `b` is itself a non-positive integer (both arguments negative integers), which previously returned `nan` instead of SciPy's `+inf`.

The reflection branch is evaluated on safe arguments where it is unused, so the added `jnp.where` does not introduce `nan` gradients for ordinary inputs (`lgamma` has infinite derivatives at its own poles).

### Behavior

| call | before | after | scipy |
|------|--------|-------|-------|
| `betaln(-1, 1)` | `nan` | `0.0000` | `0.0000` |
| `betaln(-2, 1)` | `nan` | `-0.6931` | `-0.6931` |
| `betaln(-3, 2)` | `nan` | `-1.7918` | `-1.7918` |
| `betaln(-4, 4)` | `nan` | `-1.3863` | `-1.3863` |
| `betaln(-3, -1)` | `nan` | `inf` | `inf` |

## Testing

- New regression test `testBetalnNegativeIntegers` compares against `scipy.special.betaln` for a grid of negative-integer `a` and integer `b`, under both eager execution and jit.
- Verified against SciPy over a broader mixed grid (negative integers, zero, non-integers, large values) with no regressions to the previously-correct `inf`/`-inf`/finite cases.
- Verified gradients for ordinary (positive) inputs are unchanged and match `digamma(a) - digamma(a + b)`.

### CODE REVIEW

**Comment by jakevdp** [comment]
This falls under #38651; please note that it may take a while for this to receive attention from the team.
